### PR TITLE
Move filename validation into adaptor for AWS "directories".  Add preserveFileName Support

### PR DESCRIPTION
**Pull request by mpatnode**
Sorry for mixing the two features here, but they were interdependent.   The first was to implement file validation for AWS files.    This validation method will be ignored until https://github.com/parse-community/parse-server/pull/6157 is merged.   

The end result here is that now you can specify an AWS "directory" path in your filename and it won't be rejected by a Router class that doesn't understand your filesystem.   You can now supply methods for filename validation and AWS key generation, so you can decide what's a valid filename, and how you want to deal with duplicate filenames.   Note the default behavior for the later needs to be disabled in parse-server:FilesController if you want to use directories in AWS.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2,6 +2,8 @@
 //
 // Stores Parse files in AWS S3.
 
+const FilesAdapter = require('parse-server/lib/Adapters/Files/FilesAdapter').FilesAdapter;
+const Parse = require('parse/lib/node/Parse').Parse;
```
Comment (mpatnode) [inline index.js]: I needed to bring in Parse to throw the right error codes.

File: index.js [hunk 2]
```diff
@@ -11,11 +13,12 @@ const awsCredentialsDeprecationNotice = function awsCredentialsDeprecationNotice
     'See: https://github.com/parse-server-modules/parse-server-s3-adapter#aws-credentials for details');
 };
 
-class S3Adapter {
+class S3Adapter extends FilesAdapter {
```
Comment (mpatnode) [inline index.js]: This seemed like the right thing to do, even though the parse-server PR doesn't actually change the signature.

File: spec/test.spec.js [hunk 3]
```diff
@@ -279,60 +279,141 @@ describe('S3Adapter tests', () => {
...
+    });
+  });
 
-        callback(null, {
-          Location: `https://${bucket}.s3.amazonaws.com/${Key}`,
-        });
-      }, 100),
-      deleteObject: (params, callback) => setTimeout(() => {
-        const { Key } = params;
+  function makeS3Adaptor(options) {
```
Comment (mpatnode) [inline spec/test.spec.js]: This code came in handy for more createFile requests.

File: spec/test.spec.js [hunk 4]
```diff
@@ -279,60 +279,141 @@ describe('S3Adapter tests', () => {
...
+      const fileName = 'foo/randomFileName.txt';
+      const response = s3.createFile(fileName, 'hello world', 'text/utf8').then((value) => {
+        const url = new URL(value.Location);
+        expect(url.pathname.indexOf('foo/')).toEqual(1);
+        expect(url.pathname.indexOf('random') > 13).toBe(true);
+      });
+      promises.push(response);
+    });
+
+    afterAll(() => Promise.all(promises));
```
Comment (mpatnode) [inline spec/test.spec.js]: Don't forget to do this.   Otherwise, they resolve in filesAdapterTests.testAdapter(), which breaks reporting.

**Comment by dplewis** [comment]
@acinader What do you think? This is similar to https://github.com/parse-community/parse-server-s3-adapter/issues/16. I think it fits the suggestions.

File: package.json [hunk 5]
```diff
@@ -25,7 +25,9 @@
   },
   "homepage": "https://github.com/parse-community/parse-server-s3-adapter#readme",
   "dependencies": {
-    "aws-sdk": "^2.59.0"
+    "aws-sdk": "^2.59.0",
+    "parse": "^2.8.0",
+    "parse-server": "^3.9.0"
```
Comment (dplewis) [inline package.json]: Remove the parse-server dependency?
Pin the parse dependency? (remove the ^) we should pin all dependencies as some point.

File: spec/test.spec.js [hunk 6]
```diff
@@ -279,60 +280,141 @@ describe('S3Adapter tests', () => {
...
+  });
+
+  filesAdapterTests.testAdapter('S3Adapter', makeS3Adaptor({}));
```
Comment (dplewis) [inline spec/test.spec.js]: Adaptor vs Adapter, Sorry for the nit

**Comment by yomybaby** [comment]
@dplewis Ok. I'll take a look at it tonight.

File: package.json (hunk 5 above)
Comment (mpatnode) [reply package.json]: Guess I don't understand why.   How should I deal with throwing Parse.Error?  Is there another way to get the right type declared?   Still feels wrong to not extend FilesAdapter.   Seems to me they are actually dependent...

Comment (mpatnode) [reply package.json]: OK.  Thought more about it.   I have an idea.

Comment (mpatnode) [reply package.json]: All gone.

File: spec/test.spec.js (hunk 6 above)
Comment (mpatnode) [reply spec/test.spec.js]: My bad

File: README.md [hunk 7]
```diff
@@ -169,6 +176,8 @@ var s3Options = {
   baseUrl: process.env.SPACES_BASE_URL, 
   region: process.env.SPACES_REGION,
   directAccess: true,
+  preserveFilename: "always",
+  fileNameCheck: "safe",
```
Comment (mpatnode) [inline README.md]: FWIW: I'm only using these settings in my app.   Happy to remove the others if they are at controversial.

**Comment by yomybaby** [comment]
Move filename validation out of the Router and into the FilesAdaptor(https://github.com/parse-community/parse-server/pull/6157) is great idea.

How about make this more customizable?
There is a two options instead of `preserveFilename` and `fileNameCheck`.
There are function.

like this
```
new S3Adapter({
  bucket: "my_bucket",
  generateKey: (filename) => {
    return 'origin/'+filename;
  },
  validateFilename: (filename) => {
    if (filename.length > 1024) {
       return 'Filename too long.';
     }
     return null;
  }
})
```

**Comment by mpatnode** [comment]
Thought about it,  but use my case was fairly simple.   If it's needed in the future, the validation code could simply check the type of the validate/preserveFilename parameter, and do the right thing.   I wanted to preserve the same filename check behavior from the old code, while enabling the directory capability.   Happy to remove the preserveFilename feature until someone needs it.  I'm not using it.

**Comment by mpatnode** [comment]
> How about make this more customizable?
> There is a two options instead of `preserveFilename` and `fileNameCheck`.
> There are function.

So don't know when I'll have time to get back to this.  @yomybaby Any interest in implementing that on top of my branch?

**Comment by dplewis** [comment]
@mpatnode Sorry for the late reply. Adding parse module (not parse-server) for Parse.Error should be ok here as we use it in the parse-server-push-adapter

https://github.com/parse-community/parse-server-push-adapter/blob/master/package.json#L39

**Comment by mpatnode** [comment]
> @mpatnode Sorry for the late reply. Adding parse module (not parse-server) for Parse.Error should be ok here as we use it in the parse-server-push-adapter

Hmm.  Well the interface change on the Parse side was already checked in:  https://github.com/parse-community/parse-server/pull/6166/files.   I'm reluctant to reverse this and force a new API version change in parse-server.  Do you think it's worth the trouble?

File: lib/optionsFromArguments.js [hunk 8]
```diff
@@ -90,6 +92,8 @@ const optionsFromArguments = function optionsFromArguments(args) {
   options = fromEnvironmentOrDefault(options, 'baseUrlDirect', 'S3_BASE_URL_DIRECT', false);
   options = fromEnvironmentOrDefault(options, 'signatureVersion', 'S3_SIGNATURE_VERSION', 'v4');
   options = fromEnvironmentOrDefault(options, 'globalCacheControl', 'S3_GLOBAL_CACHE_CONTROL', null);
+  options = fromOptionsDictionaryOrDefault(options, 'generateKey', null);
+  options = fromOptionsDictionaryOrDefault(options, 'validateFilename', null);
```
Comment (mpatnode) [inline lib/optionsFromArguments.js]: JS code in env vars didn't seem like a good idea to me...

File: README.md [hunk 9]
```diff
@@ -109,7 +112,16 @@ var s3Adapter = new S3Adapter('accessKey',
...
+                    globalCacheControl: 'public, max-age=86400',  // 24 hrs Cache-Control.
+                    validateFilename: (filename) => {
+                      if (filename.length > 1024) {
+                         return 'Filename too long.';
+                       }
+                       return null; // Return null on success
+                    },
+                    generateKey: (filename) => {
+                        return `${Date.now()}_${filename}`; // unique prefix for every filename
+                    }
```
Comment (mpatnode) [inline README.md]: @yomybaby This look good to you?

File: package.json [hunk 10]
```diff
@@ -25,7 +25,8 @@
   },
   "homepage": "https://github.com/parse-community/parse-server-s3-adapter#readme",
   "dependencies": {
-    "aws-sdk": "^2.59.0"
+    "aws-sdk": "^2.59.0",
+    "parse": "^2.9.1"
```
Comment (mpatnode) [inline package.json]: @dplewis 👍

File: README.md [hunk 11]
```diff
@@ -109,7 +112,19 @@ var s3Adapter = new S3Adapter('accessKey',
                     directAccess: false,
                     baseUrl: 'http://images.example.com',
                     signatureVersion: 'v4',
-                    globalCacheControl: 'public, max-age=86400'  // 24 hrs Cache-Control.
+                    globalCacheControl: 'public, max-age=86400',  // 24 hrs Cache-Control.
+                    validateFilename: (filename) => {
+                      if (filename.length > 1024) {
+                         return new Parse.Error(
+                            Parse.Error.INVALID_FILE_NAME,
```
Comment (mpatnode) [inline README.md]: @dplewis Hmm... this might be problematic?

Comment (mpatnode) [reply README.md]: No longer necessary: https://github.com/parse-community/parse-server/pull/6246

Comment (dplewis) [reply README.md]: You should `throw new Parse.Error` instead of returning

Comment (dplewis) [reply README.md]: Actually ignore that last comment this is correct.

**Comment by mpatnode** [comment]
@dplewis This is happy now: https://github.com/parse-community/parse-server/pull/6246

File: README.md (hunk 11 above)
Comment (mpatnode) [reply README.md]: @dplewis @acinader So are we good to go here?

**Pull request merged**